pygmt/clib/session.py

```diff
@@ -1430,14 +1435,25 @@ def virtualfile_from_data(self, check_kind=None, data=None, x=None, y=None, z=No
         if kind in ("file", "grid"):
             _data = (data,)
         elif kind == "vectors":
-            _data = (x, y, z)
+            _data = [np.atleast_1d(x), np.atleast_1d(y)]
+            if z is not None:
+                _data.append(np.atleast_1d(z))
+            if extra_arrays:
+                _data.extend(extra_arrays)
         elif kind == "matrix":  # turn 2D arrays into list of vectors
             try:
                 # pandas.DataFrame and xarray.Dataset types
                 _data = [array for _, array in data.items()]
             except AttributeError:
-                # Python lists, tuples, and numpy ndarray types
-                _data = np.atleast_2d(np.asanyarray(data).T)
+                try:
+                    # Just use virtualfile_from_matrix for 2D
+                    # numpy.ndarray which are not datetime (M) types
+                    assert data.ndim == 2 and not data.dtype.kind == "M"
+                    _virtualfile_from = self.virtualfile_from_matrix
+                    _data = (data,)
+                except (AssertionError, AttributeError):
+                    # Python lists, tuples, and numpy ndarray types
+                    _data = np.atleast_2d(np.asanyarray(data).T)
```

Comment:
Would appreciate advice on improving/refactoring this chunk of code. I'm half thinking whether to move some of the logic to `pygmt/helpers/utils.py`, i.e. have a new 'kind' besides file/grid/matrix/vectors:

https://github.com/GenericMappingTools/pygmt/blob/65b099a5873b4f9e76d7fccdbc8882d9b37870e3/pygmt/helpers/utils.py#L65-L72

Or we could just keep things like this as it is.

Reply:
I didn't read the codes carefully, but I think the "matrix" kind is quite confusing. 

The GMT API function `GMT_Put_Matrix()`, the PyGMT wrapper `put_matrix()` and the virtualfile function `virtualfile_from_matrix()` all require a simple 2d matrix with a single dtype (e.g., `np.float` or `np.double`). 

However, currently, data types like pandas.DataFrame are also "matrix". So we have to check the data types to choose either `virtualfile_from_vector` or `virtualfile_from_matrix`.

I agree with you that we can/shoudl add a new kind to distinguish between a GMT-compatible "matrix" and a more complicated data structure.

Reply:
How about we keep this logic as is for now to resolve #1021 (which helps with #1020), and refactor things later to have more specific 'matrix' types?

Reply:
Sounds good.